Implement experimental allow="private-token" permissions policy

sysrqb opened the pull request:
#### 2887eed67e7d43da879523f2b3f8a9b3aee2a406
<pre>
Implement experimental allow=&quot;private-token&quot; permissions policy
<a href="https://bugs.webkit.org/show_bug.cgi?id=266549">https://bugs.webkit.org/show_bug.cgi?id=266549</a>
<a href="https://rdar.apple.com/107854094">rdar://107854094</a>

Reviewed by Youenn Fablet.

This patch implements a new experimental permissions policy that provides
third-party contexts with access to using Private Tokens, as described in the
explainer [0]. If a third-party context is not granted the permission, then it
is not allowed to interact in the specified HTTP Authorization [1] flow. If the
context is given permission, then that interaction happens at a lower layer
than WebKit on Cocoa platforms.

[0] <a href="https://github.com/WebKit/explainers/tree/main/ThirdPartyPrivateTokens">https://github.com/WebKit/explainers/tree/main/ThirdPartyPrivateTokens</a>
[1] <a href="https://datatracker.ietf.org/doc/draft-ietf-privacypass-auth-scheme/">https://datatracker.ietf.org/doc/draft-ietf-privacypass-auth-scheme/</a>

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/html/FeaturePolicy.cpp:
(WebCore::policyTypeName):
(WebCore::FeaturePolicy::parse):
(WebCore::FeaturePolicy::allows const):
* Source/WebCore/html/FeaturePolicy.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setAsIsolatedCopy):
(WebCore::ResourceRequestBase::setIsPrivateTokenUsageByThirdPartyAllowed):
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::isPrivateTokenUsageByThirdPartyAllowed const):
* Source/WebCore/platform/network/cf/ResourceRequest.h:
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):
(WebCore::ResourceRequest::getResourceRequestPlatformData const):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setShouldSendPrivateTokenIPCForTesting const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startNetworkLoad):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setShouldSendPrivateTokenIPCForTesting):
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::shouldSendPrivateTokenIPCForTesting const):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setPrivateTokenIPCForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didAllowPrivateTokenUsageByThirdPartyForTesting):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::didAllowPrivateTokenUsageByThirdPartyForTesting):
(WebKit::WebsiteDataStore::setPrivateTokenIPCForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::didAllowPrivateTokenUsageByThirdPartyForTesting):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(-[NavigationDelegate websiteDataStore:didAllowPrivateTokenUsageByThirdPartyForTesting:forResourceURL:]):
(setupWebViewForPrivateTokenTests):
(TEST):

Canonical link: <a href="https://commits.webkit.org/273360@main">https://commits.webkit.org/273360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fc681263be9309740ba4e14b2c2645ac4769f82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37106 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37776 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31637 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36182 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16303 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30578 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11797 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10343 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10402 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39028 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/29758 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31861 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31661 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36422 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/34912 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10503 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8433 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34409 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12301 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/41533 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8058 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11035 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8643 "Found 231 new JSC stress test failures: microbenchmarks/array-from-derived-object-func.js.bytecode-cache, microbenchmarks/array-from-derived-object-func.js.default, microbenchmarks/array-from-derived-object-func.js.dfg-eager, microbenchmarks/array-from-derived-object-func.js.dfg-eager-no-cjit-validate, microbenchmarks/array-from-derived-object-func.js.eager-jettison-no-cjit, microbenchmarks/array-from-derived-object-func.js.no-cjit-collect-continuously, microbenchmarks/array-from-derived-object-func.js.no-llint, microbenchmarks/array-from-object-func.js.bytecode-cache, microbenchmarks/array-from-object-func.js.default, microbenchmarks/array-from-object-func.js.dfg-eager ... (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11366 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10023,6 +10023,10 @@ void WebPageProxy::didApplyLinkDecorationFiltering(const URL& originalURL, const
+void WebPageProxy::didAllowPrivateTokensForThirdPartyForTesting(bool wasAllowed, const URL& resourceURL)
+{
+    m_navigationClient->didAllowPrivateTokensForThirdPartyForTesting(wasAllowed, resourceURL);
+}
```
new line

youennf commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1648,6 +1648,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
     void didBlockLoadToKnownTracker(const URL&);
     void didApplyLinkDecorationFiltering(const URL&, const URL&);
+    void didAllowPrivateTokensForThirdPartyForTesting(bool wasAllowed, const URL& resourceURL);
```
```suggestion
    void didAllowPrivateTokensForThirdPartyForTesting(bool wasAllowed, const URL&);
```

youennf commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -361,6 +364,20 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
     if (isMainFrameNavigation)
         parameters.linkPreconnectEarlyHintsEnabled = mainFrame->settings().linkPreconnectEarlyHintsEnabled();
+    else if (RefPtr localMainFrame = dynamicDowncast<LocalFrame>(mainFrame)) {
+        if (auto* document = frame->document(); localMainFrame->document()
```
```suggestion
        if (RefPtr document = frame->document(); localMainFrame->document()
```

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5433,6 +5433,19 @@ void Internals::setTrackingPreventionEnabled(bool enable)
 
+void Internals::setSendPrivateTokenIPCForTesting(bool enabled)
+{
+    Document* document = contextDocument();
```
```suggestion
    RefPtr document = contextDocument();
```

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -5433,6 +5433,19 @@ void Internals::setTrackingPreventionEnabled(bool enable)
+    if (!document)
+        return;
+
+    Page* page = document->page();
```
```suggestion
    RefPtr page = document->page();
```

youennf commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -361,6 +364,20 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
+            std::optional<SecurityOriginData> securityOrigin;
+            if (cachedResource && cachedResource->type() == CachedResource::Type::MainResource)
+                securityOrigin = SecurityOriginData::fromURL(parameters.request.url());
+            parameters.isPermissionPolicyPrivateTokenAllowed = isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type::PrivateToken, *document, LogFeaturePolicyFailure::No, securityOrigin);
```
I would tend to do this computation at call site in CachedResourceLoader/CachedResource and add a ResourceLoaderOptions.isPrivacyTokenUsageAllowed.

The service worker integration would also be interesting to study.
If there is a call from a third party iframe that is allowed to use privacy token, it seems the intercepted request by the service worker without modification should also be allowed to use privacy token.
In that case, we need to send that information to the service worker.

annevk replied:
Do we have a request representation that goes through service workers and ends up in CFNetwork? It seems setting a bit on that that private-token access is enabled for this request should suffice.

(Now once we have site isolation we might have to think harder about this as we could/should presumably constrain cross-site processes better when they don't have certain permissions.)

sysrqb replied:
Thanks for these reviews! I moved the computation into FrameLoader and added the result into ResourceRequest. Using ResourceRequest seems to transparently pass the information through a ServiceWorker, too.

youennf commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -361,6 +364,20 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
+        }
+
+        if (auto* page = frame && frame->page() ? WebPage::fromCorePage(*frame->page()) : nullptr; page && page->shouldSendPrivateTokenIPCForTesting())
+            page->send(Messages::WebPageProxy::DidAllowPrivateTokensForThirdPartyForTesting(parameters.isPermissionPolicyPrivateTokenAllowed, parameters.request.url()));
```
I am not sure this is the right place for this.
Maybe it should be networking process that would send this information to UIProcess at the moment it triggers the network load. NetworkResourceLoader::startNetworkLoad for instance.

youennf commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -361,6 +364,20 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
+
+            std::optional<SecurityOriginData> securityOrigin;
+            if (cachedResource && cachedResource->type() == CachedResource::Type::MainResource)
+                securityOrigin = SecurityOriginData::fromURL(parameters.request.url());
```
Permission policy is usually checking the origin of the context doing the work be it capture or networking. I do not see why privacy tokens are special here and why the request's origin is used here.

sysrqb replied:
The behavior was surprising to me without this change, but after thinking about this more I agree that the current behavior is correct. Permission was denied when we requested the main resource for a new cross-origin iframe without specifying the parent frame's origin in the allowlist. This is because the `document`'s origin at the time that we decide if the feature is allowed is the parent document's origin, and `src` is computed from the iframe's src. I was expecting it to work because the top document is allowed all permissions by default, therefore an iframe that it creates should be allowed, too. But this is wrong, in retrospect.

youennf commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -1001,6 +1001,8 @@ typedef (FetchRequest or FetchResponse) FetchObject;
     DOMString resourceLoadStatisticsForURL(DOMURL url);
     undefined setTrackingPreventionEnabled(boolean enable);
 
+    undefined setSendPrivateTokenIPCForTesting(boolean enabled);
```
How about using a UnifedWebPreference/Settings.
This way, InternalsSettings can be automatically generated and used.

youennf commented on Source/WebCore/html/FeaturePolicy.cpp:
```diff
@@ -77,15 +77,21 @@ static const char* policyTypeName(FeaturePolicy::Type type)
 }
 
-bool isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type type, const Document& document, LogFeaturePolicyFailure logFailure)
+bool isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type type, const Document& document, LogFeaturePolicyFailure logFailure, std::optional<SecurityOriginData> securityOrigin)
```
I am not exactly sure why we need a securityOrigin here, the one of the document and its ancestors should be sufficient in theory.

annevk commented on Source/WebCore/html/FeaturePolicy.cpp:
```diff
@@ -298,6 +307,12 @@ FeaturePolicy FeaturePolicy::parse(Document& document, const HTMLIFrameElement&
+        }
     }
 
     // By default, camera, microphone, speaker-selection, display-capture, fullscreen, xr-spatial-tracking, screen-wake-lock, and web-share policy is 'self'.
```
Should we update this comment for accuracy? Or perhaps we can document it through code instead by having

```auto self = document.securityOrigin().data()```

sysrqb replied:
Good idea, I'll clean up the other entries in a followup.

annevk reviewed (commented):
I think @youennf is correct on the origin checks. In particular, a top-level document is not constrained in any way. So a nested document with the `private-token` Permission Policy should not be constrained either.

youennf commented on Source/WebKit/NetworkProcess/NetworkLoadParameters.h:
```diff
@@ -96,6 +97,7 @@ class NetworkLoadParameters {
     bool hadMainFrameMainResourcePrivateRelayed { false };
     bool allowPrivacyProxy { true };
     OptionSet<WebCore::AdvancedPrivacyProtections> advancedPrivacyProtections;
+    bool isPermissionPolicyPrivateTokenAllowed { false };
```
This should not be needed, either the request has the flag or it does not have it. Or am I missing something?

youennf commented on Source/WebCore/html/FeaturePolicy.cpp:
```diff
@@ -306,9 +310,16 @@ FeaturePolicy FeaturePolicy::parse(Document& document, const HTMLIFrameElement&
     }
 
     // By default, camera, microphone, speaker-selection, display-capture, fullscreen, xr-spatial-tracking, screen-wake-lock, and web-share policy is 'self'.
+    auto& defaultAllowlistSelf = document.securityOrigin().data();
```
I am not sure we like temp variables like this these days. s/auto&/auto/
In which case, we should do this change for all policies where it makes sense.

sysrqb replied:
I didn't want to make this patch more confusing by changing all of the policies. I was going to make that change in a follow up patch.

sysrqb replied:
Actually, I'll follow the current style for this patch, and I'll make this change in a separate PR (along with some other changes in this file).

youennf commented on Source/WebCore/platform/network/ResourceRequestBase.cpp:
```diff
@@ -672,6 +673,13 @@ void ResourceRequestBase::setDidFilterLinkDecoration(bool didFilterLinkDecoratio
+void ResourceRequestBase::setIsPrivateTokenUsageByThirdPartyAllowed(bool isPrivateTokenUsageByThirdPartyAllowed)
+{
+    if (m_requestData.m_isPrivateTokenUsageByThirdPartyAllowed == isPrivateTokenUsageByThirdPartyAllowed)
+        return;
```
I know this follows the above code style, but I am not sure it is worth doing this check.

youennf commented on Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
```diff
@@ -70,6 +70,7 @@ NetworkResourceLoadParameters::NetworkResourceLoadParameters(
     , std::optional<UserContentControllerIdentifier> userContentControllerIdentifier
 #endif
     , bool linkPreconnectEarlyHintsEnabled
+    , bool shouldSendPrivateTokenIPCForTesting
```
Does it make sense to send it for all requests?
Given we are using test APIs, should we instead have NetworkProcessProxy instructs its NetworkProcess to send the private token for testing?

See for instance the Testing messages in NetworkProcess.messages.in

youennf commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -253,6 +253,11 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
         [mutableRequest _setWebSearchContent:YES];
 
+#if HAVE(ALLOW_PRIVATE_ACCESS_TOKENS_FOR_THIRD_PARTY)
+    if ([mutableRequest respondsToSelector:@selector(_setAllowPrivateAccessTokensForThirdParty:)] && parameters.isPermissionPolicyPrivateTokenAllowed)
```
Shouldn't it be:
```suggestion
    if ([mutableRequest respondsToSelector:@selector(_setAllowPrivateAccessTokensForThirdParty:)] && parameters.request.isPrivateTokenUsageByThirdPartyAllowed())
```

youennf commented on Source/WebCore/html/FeaturePolicy.cpp:
```diff
@@ -88,6 +90,7 @@ bool isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type type, cons
 {
     auto& topDocument = document.topDocument();
     auto* ancestorDocument = &document;
+
```
Not needed

youennf commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5234,6 +5234,34 @@ PrivateClickMeasurementFraudPreventionEnabled:
       "HAVE(RSA_BSSA)": true
       default: false
 
+PrivateTokenIPCForTestingEnabled:
```
I think we can fully remove this flag and directly add a WKWebSiteDataStore private API to set this value.
We can simply do like _setResourceLoadStatisticsTimeAdvanceForTesting and, when calling the setter for this flag, send the IPC message to network process.

youennf commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1777,6 +1777,12 @@ void WebsiteDataStore::clearResourceLoadStatisticsInWebProcesses(CompletionHandl
 
+void WebsiteDataStore::didAllowPrivateTokenUsageByThirdPartyForTesting(bool wasAllowed, URL&& resourceURL)
+{
+    for (auto& page : m_pages)
```
We are sending the delegate to all pages, but at most one page is responsible for this load.
Either we use a website data store delegate or we pass in the IPC message a WebPageProxyIdentifier so that we directly hop from NetworkProcessProxy to WebPageProxy.
Given we want to test service worker, a website data store delegate might be better for now.

sysrqb replied:
Yes, I'll investigate the website datastore delegate. Initially I used the WebPageProxyIdentifier, but that failed when the resource load is initiated by the service worker, and that is why I used this method.

youennf commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1805,6 +1811,26 @@ void WebsiteDataStore::updateBlobRegistryPartitioningState()
     }
 }
 
+bool WebsiteDataStore::shouldSendPrivateTokenIPCForTesting() const
```
If we directly send the IPC message as part of the data store testing SPI, we would not need this.

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1991,3 +1994,354 @@ HTTPServer httpServer({
+"</script>"_s;
+
+    static constexpr auto serviceWorkerBytes =
+"self.addEventListener(\"fetch\", (event) => {"
```
Can we do a fetch in "install" event handler and verify request.isPrivateTokenUsageByThirdPartyAllowed is false.

I would also tend to move the service worker specific tests in its own test.
This might make maintenance of the test easier.

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1991,3 +1994,354 @@ HTTPServer httpServer({
+        return String(headerValueStart, hostHeaderEnd - headerValueStart);
+    };
+
+    bool didDecideServerWorkerRequest { false };
```
```suggestion
    bool didDecideServiceWorkerRequest { false };
```

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1991,3 +1994,354 @@ HTTPServer httpServer({
+    didDecideServerWorkerRequest = false;
+    didReceiveAllowPrivateToken = false;
+
+    // iframe with allowslist="https://site1.example https://site4.example"
```
Maybe we should split those tests as well in their own tests, to ease future maintenance?
This would require introducing a helper routine to share common  test infra code.

youennf reviewed (commented):
Sorry for the iterative review. This patch LGTM with below changes:
- I would tend to further simply how the testing flag is sent to network process (see below suggestion).
- As of the WKWebView delegate, given you are testing service workers, I think the latter might be better (there is no WKWebView object for a service worker).

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1991,3 +2017,403 @@ HTTPServer httpServer({
+        const char* hostHeaderEnd = strnstr(headerValueStart, "\r\n", request.size() - (headerValueStart - request.data()));
+        if (!hostHeaderEnd)
+            return ""_str;
+        return String(headerValueStart, hostHeaderEnd - headerValueStart);
```
We usually use StringView for string manipulation, find...
Also String::fromUTF8 might be better.

sysrqb replied:
Sure, I can change this. I mostly copied this from the existing implementation of HTTPServer::parsePath

sysrqb committed:
Implement experimental allow="private-token" permissions policy
https://bugs.webkit.org/show_bug.cgi?id=266549
rdar://107854094

Reviewed by Youenn Fablet.

This patch implements a new experimental permissions policy that provides
third-party contexts with access to using Private Tokens, as described in the
explainer [0]. If a third-party context is not granted the permission, then it
is not allowed to interact in the specified HTTP Authorization [1] flow. If the
context is given permission, then that interaction happens at a lower layer
than WebKit on Cocoa platforms.

[0] https://github.com/WebKit/explainers/tree/main/ThirdPartyPrivateTokens
[1] https://datatracker.ietf.org/doc/draft-ietf-privacypass-auth-scheme/

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/html/FeaturePolicy.cpp:
(WebCore::policyTypeName):
(WebCore::FeaturePolicy::parse):
(WebCore::FeaturePolicy::allows const):
* Source/WebCore/html/FeaturePolicy.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setAsIsolatedCopy):
(WebCore::ResourceRequestBase::setIsPrivateTokenUsageByThirdPartyAllowed):
* Source/WebCore/platform/network/ResourceRequestBase.h:
(WebCore::ResourceRequestBase::RequestData::RequestData):
(WebCore::ResourceRequestBase::isPrivateTokenUsageByThirdPartyAllowed const):
* Source/WebCore/platform/network/cf/ResourceRequest.h:
* Source/WebCore/platform/network/cocoa/ResourceRequestCocoa.mm:
(WebCore::ResourceRequest::ResourceRequest):
(WebCore::ResourceRequest::getResourceRequestPlatformData const):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::setShouldSendPrivateTokenIPCForTesting const):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::startNetworkLoad):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setShouldSendPrivateTokenIPCForTesting):
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::shouldSendPrivateTokenIPCForTesting const):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setPrivateTokenIPCForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didAllowPrivateTokenUsageByThirdPartyForTesting):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::didAllowPrivateTokenUsageByThirdPartyForTesting):
(WebKit::WebsiteDataStore::setPrivateTokenIPCForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::didAllowPrivateTokenUsageByThirdPartyForTesting):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(-[NavigationDelegate websiteDataStore:didAllowPrivateTokenUsageByThirdPartyForTesting:forResourceURL:]):
(setupWebViewForPrivateTokenTests):
(TEST):

Canonical link: https://commits.webkit.org/273360@main
